samples/api/hello_triangle/hello_triangle.cpp

```diff
@@ -277,40 +277,42 @@ void HelloTriangle::init_device(Context &                        context,
 	std::vector<VkPhysicalDevice> gpus(gpu_count);
 	VK_CHECK(vkEnumeratePhysicalDevices(context.instance, &gpu_count, gpus.data()));
 
-	// Select the first GPU we find in the system.
-	context.gpu = gpus.front();
+	for(size_t i = 0;i < gpu_count && (context.graphics_queue_index < 0);i++) {
 
-	uint32_t queue_family_count;
-	vkGetPhysicalDeviceQueueFamilyProperties(context.gpu, &queue_family_count, nullptr);
+		context.gpu = gpus[i];
 
-	if (queue_family_count < 1)
-	{
-		throw std::runtime_error("No queue family found.");
-	}
+		uint32_t queue_family_count;
+		vkGetPhysicalDeviceQueueFamilyProperties(context.gpu, &queue_family_count, nullptr);
 
-	std::vector<VkQueueFamilyProperties> queue_family_properties(queue_family_count);
-	vkGetPhysicalDeviceQueueFamilyProperties(context.gpu, &queue_family_count, queue_family_properties.data());
+		if (queue_family_count < 1)
+		{
+			throw std::runtime_error("No queue family found.");
+		}
 
-	uint32_t device_extension_count;
-	VK_CHECK(vkEnumerateDeviceExtensionProperties(context.gpu, nullptr, &device_extension_count, nullptr));
-	std::vector<VkExtensionProperties> device_extensions(device_extension_count);
-	VK_CHECK(vkEnumerateDeviceExtensionProperties(context.gpu, nullptr, &device_extension_count, device_extensions.data()));
+		std::vector<VkQueueFamilyProperties> queue_family_properties(queue_family_count);
+		vkGetPhysicalDeviceQueueFamilyProperties(context.gpu, &queue_family_count, queue_family_properties.data());
 
-	if (!validate_extensions(required_device_extensions, device_extensions))
-	{
-		throw std::runtime_error("Required device extensions are missing, will try without.");
-	}
+		uint32_t device_extension_count;
+		VK_CHECK(vkEnumerateDeviceExtensionProperties(context.gpu, nullptr, &device_extension_count, nullptr));
+		std::vector<VkExtensionProperties> device_extensions(device_extension_count);
+		VK_CHECK(vkEnumerateDeviceExtensionProperties(context.gpu, nullptr, &device_extension_count, device_extensions.data()));
 
-	for (uint32_t i = 0; i < queue_family_count; i++)
-	{
-		VkBool32 supports_present;
-		vkGetPhysicalDeviceSurfaceSupportKHR(context.gpu, i, context.surface, &supports_present);
+		if (!validate_extensions(required_device_extensions, device_extensions))
+		{
+			throw std::runtime_error("Required device extensions are missing, will try without.");
```

Comment:
I think this might need to be rethought a little bit.  As this code now no longer looks at only the first device, it probably makes sense to only throw if no device has all requested extensions in support.

Reply:
Fair. I do not want to make the hello app too complex, but I can definitely reorder this in a way that makes more sense.

Reply:
> Required device extensions are missing, will try without.

Assuming that is accurate, it does not make any sense to me to use this to influence the gpu selection for this specific sample, but still worth a check. Extension support might be something to consider for the framework gpu selection logic, though.